srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,201 @@
+		cds.gen.reviewservice.Reviews inserted = res.single(cds.gen.reviewservice.Reviews.class);
+
+		messages.success(MessageKeys.REVIEW_ADDED);
+
+		context.setResult(Struct.access(inserted).as(Reviews.class));
+	}
+
+	/**
+	 * Recalculates and sets the book rating after a new review for the given book.
+	 *
+	 * @param context {@link ReviewContext}
+	 */
+	@After(entity = Books_.CDS_NAME)
+	public void afterAddReview(AddReviewContext context) {
+		bookRatingService.setBookRating(context.getResult().getBookId());
 	}
 
 	@After(event = CdsService.EVENT_READ)
 	public void discountBooks(Stream<Books> books) {
-		books.filter(b -> b.getTitle() != null).forEach(
-			b -> {
-				loadStockIfNotSet(b);
-				discountBooksWithMoreThan111Stock(b);
+		books.filter(b -> b.getTitle() != null).forEach(b -> {
+			loadStockIfNotSet(b);
+			discountBooksWithMoreThan111Stock(b);
+		});
+	}
+
+	@After
+	public void setIsReviewable(CdsReadEventContext context, List<Books> books) {
+
+		String user = context.getUserInfo().getName();
+		List<String> bookIds = books.stream().map(b -> b.getId()).collect(Collectors.toList());
+
+		Select<Reviews_> query = Select.from(CatalogService_.BOOKS, b -> b.filter(b.ID().in(bookIds)).reviews())
+				.where(r -> r.reviewer().eq(user));
+
+		Set<String> reviewedBooks = db.run(query).streamOf(Reviews.class).map(Reviews::getBookId)
+				.collect(Collectors.toSet());
+
+		for (Books book : books) {
+			if (reviewedBooks.contains(book.getId())) {
+				book.setIsReviewable(false);
 			}
-		);
+		}
+	}
+
+	@On
+	public void onSubmitOrder(SubmitOrderContext context) {
+		Integer amount = context.getAmount();
+		String bookId = context.getBook();
+
+		Optional<Books> book = db.run(Select.from(BOOKS).columns(Books_::stock).byId(bookId)).first(Books.class);
+
+		book.orElseThrow(() -> new ServiceException(ErrorStatuses.NOT_FOUND, MessageKeys.BOOK_MISSING)
+				.messageTarget(Books_.class, b -> b.ID()));
+
+		Integer stock = book.map(Books::getStock).get();
+
+		if (stock >= amount) {
+			db.run(Update.entity(BOOKS).byId(bookId).data(Books.STOCK, stock -= amount));
+
+			ReturnCatalogServiceSubmitOrder result = ReturnCatalogServiceSubmitOrder.create();
+			result.setStock(stock);
+
+			context.setResult(result);
+			context.setCompleted();
+		} else {
+			throw new ServiceException(ErrorStatuses.CONFLICT, String.format("%d exceeds stock for book", amount));
+		}
 	}
 
 	private void discountBooksWithMoreThan111Stock(Books b) {
-		if(b.getStock() > 111) {
+		if (b.getStock() != null && b.getStock() > 111) {
 			b.setTitle(String.format("%s -- 11%% discount", b.getTitle()));
 		}
 	}
 
 	private void loadStockIfNotSet(Books b) {
-		if (b.getStock() == null) {
-			b.setStock(db.run(Select.from(BOOKS).byId(b.getId()).columns(Books_::stock)).single(Books.class).getStock());
+		if (b.getId() != null && b.getStock() == null) {
+			b.setStock(
+					db.run(Select.from(BOOKS).byId(b.getId()).columns(Books_::stock)).single(Books.class).getStock());
```

Comment:
Can we avoid this reformatting?

Reply:
Both the Eclipse and the VSCode formatter produce this... But I'll revert it manually.